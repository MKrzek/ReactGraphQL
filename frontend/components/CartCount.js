import PropTypes from 'prop-types';
import styled from 'styled-components';
import React from 'react';

const Dot = styled.div`
  background: ${props => props.theme.green};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => <Dot>{count}</Dot>;

export default CartCount;
