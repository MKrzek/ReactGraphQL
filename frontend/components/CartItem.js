import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li``;

const CartItem = ({ cartItem }) => {
  const {
    quantity,
    item: { price, image, title },
  } = cartItem;
  return (
    <CartItemStyles>
      <img width="100" src={image} alt={title} />
      <div className="cart-item-details">
        <h3>{title}</h3>
        <p>
          {formatMoney(price * quantity)}
          {' - '}
          <em>
            {quantity} x {formatMoney(price)} each
          </em>
        </p>
      </div>
    </CartItemStyles>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired,
};

export default CartItem;
