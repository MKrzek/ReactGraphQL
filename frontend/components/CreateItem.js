import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $image: String
    $largeImage: String
    $price: Int!
  ) {
    createItem(
      title: $title
      description: $description
      image: $image
      largeImage: $largeImage
      price: $price
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');
    const res = await fetch(
      'https://api.cloudinary.com/v1_1/dmss88pfo/image/upload',
      {
        method: 'POST',
        body: data,
      }
    );
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      // eslint-disable-next-line react/no-unused-state
      largeImage: file.eager[0].secure_url,
    });
  };

  render() {
    const { image, title, description, price } = this.state;
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              e.preventDefault();
              const res = await createItem();
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              });
            }}
          >
            <Error error={error} />

            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  required
                  placeholder="Upload an image"
                  onChange={this.uploadFile}
                />
                {image && <img width="200" src={image} alt="productImage" />}
              </label>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  required
                  placeholder="Title"
                  value={title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  required
                  placeholder="Price"
                  value={price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  type="text"
                  id="description"
                  name="description"
                  required
                  placeholder="Enter a  description"
                  value={description}
                  onChange={this.handleChange}
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export { CREATE_ITEM_MUTATION };
export default CreateItem;
