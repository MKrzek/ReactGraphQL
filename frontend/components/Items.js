/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from './Item';
import Pagination from './Pagination';
import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0) {
    items(skip: $skip) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;
const Center = styled.div`
  text-align: center;
`;
const ItemList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;

export default class Items extends Component {
  render() {
    const { page } = this.props;
    return (
      <Center>
        <Pagination page={page} />
        <Query
          query={ALL_ITEMS_QUERY}
          fetchPolicy="network-only"
          variables={{ skip: page * perPage - perPage }}
        >
          {({ data, loading, error }) => {
            if (loading) return <div>Loading...</div>;
            if (error) return <div>{error.message}</div>;

            return (
              <ItemList>
                {data.items.map(item => (
                  <Item key={item.id} item={item} />
                ))}
              </ItemList>
            );
          }}
        </Query>
        <Pagination page={page} />
      </Center>
    );
  }
}
