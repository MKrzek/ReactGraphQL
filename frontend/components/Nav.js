import React from "react";
import NavStyles from "./styles/NavStyles";
import Link from "next/link";

const Nav = () => {
  return (
    <NavStyles>
      <Link href="/items">
        <a>shop</a>
      </Link>
      <Link href="/sell">
        <a>sell</a>
      </Link>
      <Link href="/signup">
        <a>signup</a>
      </Link>
      <Link href="/orders">
        <a>orders</a>
      </Link>
      <Link href="/me">
        <a>Account</a>
      </Link>
    </NavStyles>
  );
};

export default Nav;
