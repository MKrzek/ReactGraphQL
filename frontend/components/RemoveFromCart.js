import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

const RemoveButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.green};
    cursor: pointer;
  }
`;

class RemoveFromCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  };

  render() {
    const { id } = this.props;
    console.log('IDDDD-item', id);
    return (
      <Mutation mutation={REMOVE_FROM_CART_MUTATION} variables={{ id }}>
        {(removeFromCart, { loading, error }) => (
          <RemoveButton
            disabled={loading}
            onClick={() => removeFromCart().catch(err => alert(err.message))}
            title="Delete Item"
          >
            x
          </RemoveButton>
        )}
      </Mutation>
    );
  }
}

export default RemoveFromCart;
