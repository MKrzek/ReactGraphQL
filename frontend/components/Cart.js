/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { adopt } from 'react-adopt';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import CartItem from './CartItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;
export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
});

const Cart = () => (
  <Composed>
    {({
      user: {
        data: { me },
      },
      toggleCart,
      localState,
    }) => {
      if (!me) return null;
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton onClick={toggleCart} title="close">
              Close
            </CloseButton>
            <Supreme>{me.name}'s cart</Supreme>
            <p>
              You have {me.cart.reduce((prev, curr) => prev + curr.quantity, 0)}{' '}
              item
              {me.cart.length > 1 && 's'} in you cart
            </p>
          </header>
          <ul>
            {me.cart.map(item => (
              <CartItem key={item.id} cartItem={item} />
            ))}
          </ul>
          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <SickButton>Checkout</SickButton>
          </footer>
        </CartStyles>
      );
    }}
  </Composed>
);

export default Cart;
