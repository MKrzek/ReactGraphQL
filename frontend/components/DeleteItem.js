/* eslint-disable prettier/prettier */
/* eslint-disable no-restricted-globals */
/* eslint-disable react/prop-types */
/* eslint-disable react/button-has-type */
import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
// eslint-disable-next-line import/no-cycle
import {ALL_ITEMS_QUERY} from './Items'


const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

export default class DeleteItem extends Component {
  update=(cache, payload)=>{
    // update cache to match the server
    const data= cache.readQuery({query:ALL_ITEMS_QUERY})
    console.log('DTAAAA', data, payload)
    data.items= data.items.filter(item=> item.id !==payload.data.deleteItem.id)
    cache.writeQuery({query:ALL_ITEMS_QUERY, data})
  }

  render() {
    const { children, id } = this.props;
    return (
      <Mutation
      mutation={DELETE_ITEM_MUTATION}
      variables={{id}}
      update={this.update}
      >
        {(deleteItem, { error }) => (
          <button
            onClick={() => {
              if (confirm('Are you sure you want to delete this item?')) {
                deleteItem()
              }
            }}
          >
            {children}
          </button>
        )}
      </Mutation>
    );
  }
}
