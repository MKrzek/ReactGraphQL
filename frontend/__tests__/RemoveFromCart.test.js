import { mount } from 'enzyme';
import wait from 'waait';
import { ApolloConsumer } from 'react-apollo';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import RemoveFromCart, {
  REMOVE_FROM_CART_MUTATION,
} from '../components/RemoveFromCart';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

global.alert = console.log;

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem({ id: 'aaa111' })],
        },
      },
    },
  },
  {
    request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: 'aaa111' } },
    result: {
      data: {
        removeFromCart: {
          __typename: 'CartItem',
          id: 'aaa111',
        },
      },
    },
  },
];

describe('<RemoveFromCart/>', () => {
  it('renders and matches the snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <RemoveFromCart id="aaa111" />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
  });

  it('removed item from cart', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <RemoveFromCart id="aaa111" />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );

    const res = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(res.data.me.cart).toHaveLength(1);
    expect(res.data.me.cart[0].item.price).toBe(5000);
    wrapper.find('button').simulate('click');
    await wait(50);
    const res2 = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(res2.data.me.cart).toHaveLength(0);
  });
});
