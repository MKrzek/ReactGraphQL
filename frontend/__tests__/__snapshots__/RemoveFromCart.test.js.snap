// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart/> renders and matches the snapshot 1`] = `
<button
  className="RemoveFromCart__RemoveButton-wkyl42-0 iijiOR"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  x
</button>
`;
